driver-core/src/main/java/com/datastax/driver/core/Clock.java

```diff
@@ -15,25 +15,121 @@
+ * If LibC is available through JNR, and if the system property {@code com.datastax.driver.USE_NATIVE_CLOCK} is set to {@code true}
+ * (which is the default value), then {@link NativeClock} is returned, otherwise {@link SystemClock} is returned.
+ */
+class ClockFactory {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ClockFactory.class);
+
+    private static final String USE_NATIVE_CLOCK_SYSTEM_PROPERTY = "com.datastax.driver.USE_NATIVE_CLOCK";
+
+    static Clock newInstance() {
+        if (Native.isLibCLoaded() && SystemProperties.getBoolean(USE_NATIVE_CLOCK_SYSTEM_PROPERTY, true)) {
+            LOGGER.info("Using native clock to generate timestamps.");
+            return new NativeClock();
+        } else {
+            LOGGER.info("Using java.lang.System clock to generate timestamps.");
+            return new SystemClock();
+        }
+    }
+
 }
 
 /**
- * Default implementation of a clock that delegate its calls to the system clock.
+ * Default implementation of a clock that delegates its calls to the system clock.
+ *
+ * @see System#currentTimeMillis()
  */
 class SystemClock implements Clock {
+
     @Override
-    public long currentTime() {
-        return System.currentTimeMillis();
+    public long currentTimeMicros() {
+        return System.currentTimeMillis() * 1000;
     }
-}
\ No newline at end of file
+
+}
+
+/**
+ * Provides the current time with microseconds precision with some reasonable accuracy through
+ * the use of {@link Native#currentTimeMicros()}.
+ * <p/>
+ * Because calling JNR methods is slightly expensive,
+ * we only call it once per second and add the number of nanoseconds since the last call
+ * to get the current time, which is good enough an accuracy for our purpose (see CASSANDRA-6106).
+ * <p/>
+ * This reduces the cost of the call to {@link NativeClock#currentTimeMicros()} to levels comparable
+ * to those of a call to {@link System#currentTimeMillis()}.
+ */
+class NativeClock implements Clock {
+
+    private static final long ONE_SECOND_NS = NANOSECONDS.convert(1, SECONDS);
+    private static final long ONE_MILLISECOND_NS = NANOSECONDS.convert(1, MILLISECONDS);
+
+    /**
+     * Records a time in micros along with the System.nanoTime() value at the time the
+     * time is fetched.
+     */
+    private static class FetchedTime {
+
+        private final long timeInMicros;
+        private final long nanoTimeAtCheck;
+
+        private FetchedTime(long timeInMicros, long nanoTimeAtCheck) {
+            this.timeInMicros = timeInMicros;
+            this.nanoTimeAtCheck = nanoTimeAtCheck;
+        }
+    }
+
+    private final AtomicReference<FetchedTime> lastFetchedTime = new AtomicReference<FetchedTime>(fetchTimeMicros());
+
+    @Override
+    public long currentTimeMicros() {
+        FetchedTime spec = lastFetchedTime.get();
+        long curNano = System.nanoTime();
+        if (curNano > spec.nanoTimeAtCheck + ONE_SECOND_NS) {
+            spec = fetchTimeMicros();
```

Comment:
> You mean, create an empty FetchedTime with zeroes ? I think that could lead to a situation where one thread CAS a new (empty) fetched time, then another thread reads it before the first thread gets a chance to fill in the fields.

Good point, I hadn't even really considered that.  That will definitely be a problem.

> lastFetchedTime.compareAndSet(spec, spec = fetchTimeMicros());

Wouldn't that not really provide much value since it requires evaluation of fetchTimeMicros()? 

> Also I think we could turn NativeClock and SystemClock into singletons.

:+1: